sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java

```diff
@@ -1669,6 +1652,77 @@ public boolean apply(@Nonnull String input) {
     return convertedOptions;
   }
 
+  /**
+   * Returns true if the given type is a SIMPLE_TYPES, enum or any of these types in a
+   * parameterized ValueProvider.
+   */
+  private static boolean isSimpleType(Class<?> type, JavaType genericType) {
+    Class<?> unwrappedType = type.equals(ValueProvider.class)
+        ? genericType.containedType(0).getRawClass() : type;
+    return SIMPLE_TYPES.contains(unwrappedType) || unwrappedType.isEnum();
+  }
+
+  /**
+   * Returns true if the given type is a any Array or Collection of SIMPLE_TYPES or enum, or
+   * any of these types in a parameterized ValueProvider.
+   */
+  private static boolean isCollectionOrArrayOfAllowedTypes(Class<?> type, JavaType genericType) {
+    Class<?> containerType = type.equals(ValueProvider.class)
+        ? genericType.containedType(0).getRawClass() : type;
+
+    // Check if it is an array of simple types or enum.
+    if (containerType.isArray() && (SIMPLE_TYPES.contains(containerType.getComponentType())
+        || containerType.getComponentType().isEnum())) {
+        return true;
+    }
+    // Check if it is Collection of simple types or enum.
+    if (Collection.class.isAssignableFrom(containerType)) {
+      JavaType innerType = type.equals(ValueProvider.class)
```

Comment:
Add a comment here reflecting why you need to go to the contained type twice when using the ValueProvider if it can't be improved to not differentiate between the two.